std/algorithm/iteration.d

```diff
@@ -918,187 +883,106 @@ template each(alias fun = "a")
     import std.traits : Parameters;
     import std.typecons : Flag, Yes, No;
 
-private:
-    alias BinaryArgs = AliasSeq!(fun, "i", "a");
-
-    enum isRangeUnaryIterable(R) =
-        is(typeof(unaryFun!fun(R.init.front)));
-
-    enum isRangeBinaryIterable(R) =
-        is(typeof(binaryFun!BinaryArgs(0, R.init.front)));
-
-    enum isRangeIterable(R) =
-        isInputRange!R &&
-        (isRangeUnaryIterable!R || isRangeBinaryIterable!R);
-
-    enum isForeachUnaryIterable(R) =
-        is(typeof((R r) {
-            foreach (ref a; r)
-                cast(void) unaryFun!fun(a);
-        }));
+    private alias BinaryArgs = AliasSeq!(fun, "i", "a");
 
-    enum isForeachUnaryWithIndexIterable(R) =
-        is(typeof((R r) {
-            foreach (i, ref a; r)
-                cast(void) binaryFun!BinaryArgs(i, a);
-        }));
-
-    enum isForeachBinaryIterable(R) =
-        is(typeof((R r) {
-            foreach (ref a, ref b; r)
-                cast(void) binaryFun!fun(a, b);
-        }));
-
-    enum isForeachIterable(R) =
-        (!isForwardRange!R || isDynamicArray!R) &&
-        (isForeachUnaryIterable!R || isForeachBinaryIterable!R ||
-         isForeachUnaryWithIndexIterable!R);
-
-public:
-    /**
-    Params:
-        r = range or iterable over which each iterates
-     */
-    Flag!"each" each(Range)(Range r)
-    if (!isForeachIterable!Range && (
-        isRangeIterable!Range ||
-        __traits(compiles, typeof(r.front).length)))
+    public Flag!"each" each(Range)(auto ref Range range)
+    if (isInputRange!Range || isStaticArray!Range || hasMember!(Range, "opApply"))
```

Comment:
This is a simpler condition (albeit incomplete because it doesn't verify the function). Also the entire implementation is a lot shorter.

Checked the speed and memory consumed - this version is about as fast and eats 0.1% less memory.